controllers/customtasks_controller.go

```diff
@@ -0,0 +1,270 @@
+//+kubebuilder:rbac:groups=tekton.dev,resources=runs,verbs=get;list;watch
+//+kubebuilder:rbac:groups=tekton.dev,resources=runs/status,verbs=update;patch
+//+kubebuilder:rbac:groups=tekton.dev,resources=runs/finalizers,verbs=update;patch
+
+// generateBuildRun generates a BuildRun instance owned by the informed Tekton Run object, the
+// BuildRun name is randomly generated using the Run's name as base.
+func (r *CustomTasksReconciler) generateBuildRun(
+	ctx context.Context,
+	run *tknv1alpha1.Run,
+) (*v1alpha1.BuildRun, error) {
+	br := v1alpha1.BuildRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: run.GetNamespace(),
+			Name:      names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", run.Name)),
+			Labels: map[string]string{
+				filter.OwnedByTektonRun: run.Name,
+			},
+		},
+		Spec: v1alpha1.BuildRunSpec{
+			BuildRef: &v1alpha1.BuildRef{
+				APIVersion: &run.Spec.Ref.APIVersion,
+				Name:       run.Spec.Ref.Name,
+			},
+			ParamValues: filter.TektonRunParamsToShipwrightParamValues(run),
+			Timeout:     run.Spec.Timeout,
+		},
+	}
+	err := controllerutil.SetControllerReference(run, &br, r.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	return &br, nil
+}
+
+// reflectBuildRunStatusOnTektonRun reflects the BuildRun status on the Run instance.
+func (r *CustomTasksReconciler) reflectBuildRunStatusOnTektonRun(
+	logger logr.Logger,
+	run *tknv1alpha1.Run,
+	br *v1alpha1.BuildRun,
+) {
+	if br.Status.CompletionTime != nil {
+		run.Status.CompletionTime = br.Status.CompletionTime
+	}
+	if run.Status.Conditions == nil {
+		run.Status.Conditions = knativev1.Conditions{}
+	}
+
+	var lastCondition apis.Condition
+	if len(run.Status.Conditions) > 0 {
+		lastCondition = run.Status.Conditions[len(run.Status.Conditions)-1]
+	}
+
+	for _, c := range br.Status.Conditions {
+		logger.WithValues(
+			"condition-type", c.Type,
+			"condition-status", c.Status,
+			"condition-reason", c.Reason,
+			"condition-message", c.Message,
+		).V(0).Info("Reflecting BuildRun status condition on the Tekton Run owner object")
+
+		severity := apis.ConditionSeverityInfo
+		if c.Status == corev1.ConditionFalse {
+			severity = apis.ConditionSeverityError
+		}
+
+		condition := apis.Condition{
+			Type:               apis.ConditionType(string(c.Type)),
+			Status:             c.Status,
+			LastTransitionTime: apis.VolatileTime{Inner: c.LastTransitionTime},
+			Reason:             c.Reason,
+			Message:            c.Message,
+			Severity:           severity,
+		}
+
+		if !reflect.DeepEqual(lastCondition, condition) {
+			if br.Status.IsFailed(c.Type) {
+				logger.V(0).Info("Marking the Tekton Run as failed!")
+				run.Status.MarkRunFailed(string(condition.Status), condition.Message)
+			}
+			if br.IsSuccessful() {
+				logger.V(0).Info("Marking the Tekton Run as successful!")
+				run.Status.MarkRunSucceeded(string(condition.Status), condition.Message)
+			}
+			run.Status.Conditions = append(run.Status.Conditions, condition)
```

Comment:
This code is causing the list of conditions to grow while a BuildRun is running.

```
apiVersion: tekton.dev/v1alpha1
kind: Run
...
status:
  conditions:
  - lastTransitionTime: "2023-01-05T17:48:58Z"
    status: Unknown
    type: Succeeded
  - lastTransitionTime: "2023-01-05T17:48:59Z"
    message: Pending
    reason: Pending
    severity: Info
    status: Unknown
    type: Succeeded
  - lastTransitionTime: "2023-01-05T17:49:12Z"
    message: Not all Steps in the Task have finished executing
    reason: Running
    severity: Info
    status: Unknown
    type: Succeeded
  extraFields:
    buildRunName: shp-run-1672940937027-ko-build-mghj2
    buildRunNamespace: build-system
  startTime: "2023-01-05T17:48:58Z"
```

Reply:
Should it only have a single condition? I thought the conditions would be appended thus the last one represents the current state.

Reply:
For instance, checking `pod` for its conditions I found the following:

```
$ k get pods jenkins-412-7f44fdb94f-mp4t2 -o json |jq '.status.conditions'
[
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2023-01-05T10:42:03Z",
    "status": "True",
    "type": "Initialized"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2023-01-06T04:31:15Z",
    "status": "True",
    "type": "Ready"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2023-01-06T04:31:15Z",
    "status": "True",
    "type": "ContainersReady"
  },
  {
    "lastProbeTime": null,
    "lastTransitionTime": "2023-01-05T10:42:03Z",
    "status": "True",
    "type": "PodScheduled"
  }
]
```

Reply:
An array of conditions can have multiple entries. But, per `type`, there should only be one.

Reply:
That makes sense. Let me amend it.

Reply:
Please consider the latest changes, I found out a knative helper to keep the conditions based on its type.